File: glue/rust/tests/button.rs
// Button demo app
use warduino::{PinMode, PinVoltage, pin_mode, digital_write, delay, sub_interrupt, InterruptMode, mqtt_loop};

static BUTTON : u32 = 25;
static LED    : u32 = 26;

fn callback(_topic: &str, _payload: &str, _length: u32) {
    let val = digital_read(LED);
    if val == HIGH {
        digital_write(LED, PinVoltage.LOW);
    } else {
        digital_write(LED, PinVoltage.HIGH);
    }
}

#[no_mangle]
pub fn main() {
    pin_mode(BUTTON, PinMode.INPUT);
    pin_mode(LED, PinMode.OUTPUT);

    sub_interrupt(BUTTON, callback, InterruptMode.FALLING);

    loop {
        delay(1000);
        mqtt_loop();
    }
}

File: glue/assemblyscript/examples/button.ts
// config.button demo app
import {
    delay,
    digitalRead,
    digitalWrite,
    InterruptMode,
    interruptOn,
    MQTT,
    pinMode,
    PinMode,
    PinVoltage
} from "../src/index";
import {config} from "./config";

function invert(voltage: PinVoltage): PinVoltage {
    switch (voltage) {
        case PinVoltage.LOW:
            return PinVoltage.HIGH;
        case PinVoltage.HIGH:
        default:
            return PinVoltage.LOW;
    }
}

function toggleLED(_topic: string, _payload: string): void {
    // Get current status of LED
    let status = digitalRead(config.led);
    // Toggle LED
    digitalWrite(config.led, invert(status));
}

export function main(): void {
    pinMode(config.button, PinMode.INPUT);
    pinMode(config.led, PinMode.OUTPUT);

    interruptOn(config.button, InterruptMode.FALLING, toggleLED);

    while (true) {
        delay(1000);
        MQTT.loop();
    }
}

